src/goto-programs/replace_calls.cpp

```diff
@@ -0,0 +1,163 @@
+Module: Replace calls
+
+Author: Daniel Poetzl
+
+\*******************************************************************/
+
+/// \file
+/// Replace calls to given functions with calls to other given
+/// functions
+
+#include "replace_calls.h"
+
+#include <goto-programs/remove_returns.h>
+
+#include <util/base_type.h>
+#include <util/invariant.h>
+#include <util/irep.h>
+#include <util/string_utils.h>
+#include <util/suffix.h>
+
+/// Replace function calls with calls to other functions
+/// \param goto_model: goto model to modify
+/// \param replacement_list: list of strings, with each string f:g denoting a
+///   mapping between functions names; a mapping f -> g indicates that calls to
+///   f should be replaced by calls to g
+void replace_callst::operator()(
+  goto_modelt &goto_model,
+  const replacement_listt &replacement_list) const
+{
+  replacement_mapt replacement_map = parse_replacement_list(replacement_list);
+  (*this)(goto_model, replacement_map);
+}
+
+/// Replace function calls with calls to other functions
+/// \param goto_model: goto model to modify
+/// \param replacement_map: mapping between function names; a mapping f -> g
+///   indicates that calls to f should be replaced by calls to g
+void replace_callst::operator()(
+  goto_modelt &goto_model,
+  const replacement_mapt &replacement_map) const
+{
+  check_replacement_map(replacement_map);
+
+  const namespacet ns(goto_model.symbol_table);
+  goto_functionst &goto_functions = goto_model.goto_functions;
+
+  for(auto &p : goto_functions.function_map)
+  {
+    goto_functionst::goto_functiont &goto_function = p.second;
+
+    if(!goto_function.body_available())
+      continue;
+
+    goto_programt &goto_program = goto_function.body;
+
+    (*this)(goto_program, goto_functions, ns, replacement_map);
+  }
+}
+
+void replace_callst::operator()(
+  goto_programt &goto_program,
+  const goto_functionst &goto_functions,
+  const namespacet &ns,
+  const replacement_mapt &replacement_map) const
+{
+  Forall_goto_program_instructions(it, goto_program)
+  {
+    goto_programt::instructiont &ins = *it;
+
+    if(!ins.is_function_call())
+      continue;
+
+    code_function_callt &cfc = to_code_function_call(ins.code);
+    exprt &function = cfc.function();
+
+    if(function.id() != ID_symbol)
+      throw "Function pointer calls must be removed before replacing calls";
+
+    symbol_exprt &se = to_symbol_expr(function);
+    const irep_idt &id = se.get_identifier();
+
+    auto f_it1 = goto_functions.function_map.find(id);
+
+    INVARIANT(
+      f_it1 != goto_functions.function_map.end(),
+      "Called functions need to be present");
+
+    replacement_mapt::const_iterator r_it = replacement_map.find(id);
+
+    if(r_it == replacement_map.end())
+      continue;
+
+    const irep_idt &new_id = r_it->second;
+
+    auto f_it2 = goto_functions.function_map.find(new_id);
+
+    if(f_it2 == goto_functions.function_map.end())
+      throw "Replacement function " + id2string(new_id) +
```

Comment:
Use proper user input exceptions?

Reply:
Yes. PR #2703 introduces an `invalid_user_input_exceptiont`. Once it's merged we'll change most throws in here to throw that one (as part of the error handling cleanup).

Reply:
I think this should be checked when things are set up / parsed rather than during the operation; but that may just be me.